scripts/ipfs-pi-stream/process-stream.sh

```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
+
+HLS_TIME=15
+M3U8_SIZE=10
+IPFS_GATEWAY="https://ipfs.io"
+
+# Load settings
+
+# Prepare Pi Camera
+sudo modprobe bcm2835-v4l2
+sudo v4l2-ctl --set-ctrl video_bitrate=100000
+
+function startFFmpeg() {
+  while true; do
+    mv ~/ffmpeg.log ~/ffmpeg.1
+    echo 1 > ~/stream-reset
+    ffmpeg -f video4linux2 -input_format h264 -video_size 1280x720 -framerate 30 -i /dev/video0 -vcodec copy -hls_time ${HLS_TIME}  ${what}.m3u8   > ~/ffmpeg 2>&1
+    sleep 0.5
+  done
+}
+
+# Create directory for HLS content
+
+currentpath="$HOME/live"
+sudo umount $currentpath
+rm -rf $currentpath
+mkdir $currentpath
+sudo mount -t tmpfs tmpfs $currentpath
+cd $currentpath
+
+what="$(date +%Y%m%d%H%M)-LIVE"
+
+# Start ffmpeg in background
+startFFmpeg &
+
+while true; do
+  nextfile=$(ls -tr ${what}*.ts 2>/dev/null | tail -n 1)
+
+  if ! [ -z "${nextfile}" ]; then
+    # Check if the next file on the list is still being written to by ffmpeg
+    if ! [ -z "$(lsof ${nextfile} | grep ffmpeg)" ]; then
+      # Wait for file to finish writing
+      # If not finished in 45 seconds something is wrong, timeout
+      inotifywait -e close_write $nextfile -t 45
+    fi
+
+    # Grab the timecode from the m3u8 file so we can add it to the log
+    timecode=`grep -B1 ${nextfile} ${what}.m3u8 | head -n1 | awk -F : '{print $2}' | tr -d ,`
+    attempts=5
+    until [[ "${timecode}" || ${attempts} -eq 0 ]]; do
+      # Wait and retry
+      sleep 0.5
+      timecode=`grep -B1 ${nextfile} ${what}.m3u8 | head -n1 | awk -F : '{print $2}' | tr -d ,`
+      attempts=$((attempts-1))
+    done
+
+    if ! [[ "${timecode}" ]]; then
+      # Set approximate timecode
+      timecode="${HLS_TIME}.000000"
+    fi
+    
+    reset_stream="$(cat ~/stream-reset)"
+    reset_stream_marker=''
+    if [[ "${reset_stream}" -eq '1' ]]; then
+      reset_stream_marker=" #EXT-X-DISCONTINUITY"
+    fi
+    
+    echo 0 > ~/stream-reset
+    # Current UTC date for the log
+    time=`date "+%F-%H-%M-%S"`
+
+    # Add ts file to IPFS
+    ret=`ipfs add ${nextfile} 2>/dev/null > ~/tmp.txt; echo $?`
+    attempts=5
+    until [[ ${ret} -eq 0 || ${attempts} -eq 0 ]]; do
+      # Wait and retry
+      sleep 1
+      ret=`ipfs add ${nextfile} 2>/dev/null > ~/tmp.txt; echo $?`
+      attempts=$((attempts-1))
+    done
+    if [[ ${ret} -eq 0 ]]; then
+      # Update the log with the future name (hash already there)
+      echo $(cat ~/tmp.txt) ${time}.ts ${timecode}${reset_stream_marker} >> ~/process-stream.log
```

Comment:
A small thing, you could just do `ipfs add -Q ${nextfile}`, which will only give the hash. That might also be useful for the log, as `ipfs add ${nextfile} 2>/dev/null > ~/tmp.txt` means tmp.txt will read `added $hash $filename`, but maybe you already have the file name: `echo $(cat ~/tmp.txt) ****${time}.ts**** ${timecode}${reset_stream_marker} >> ~/process-stream.log`

Reply:
log is only a log of the hashes in order. Its used on line 99 to assemble the m3u8 file

If you add errors into this file you will end up with a mess.  However errors will be put in /var/log/syslog since it is daemon

Reply:
I wasn't talking about errors. I'm saying how logging the output `ipfs add ${somefile}` might be more redundant then you need, because it will say `added $hash $filename`, but maybe all you wanted to log was `$hash`.

Reply:
`ipfs add -Q ${nextfile}` is a great solution, however im not going to change it for now since this mirrors what is done in ipfs-live-stream.